### ISSUE TITLE
StackOverflow on document save when child appended to text node

### ISSUE THREAD

**Issue by jmessenger235** [zzzprojects/html-agility-pack#381]
### Description
HtmlNode.PreviousSibling sometimes returns a text node. Appending a child to the text node will cause a stack overflow on save. This looks like the same as #116. I worked around this in the production code. One or all of the following would be helpful:
- A timely error about adding children to a text node so I don't lose my marbles figuring out which code where is wrong and why the stack is overflowing.
- HtmlNode.PreviousSibling returning only a usable node and not text nodes or a PreviousHtmlNode/Sibling property with documentation on the previous sibling so it's easier to see how this works and doesn't break existing code.
- The code below working even though it doesn't really make sense given it's a text node.

### Exception
I get a stack overflow that looks like the below image on the call stack.
![image](https://user-images.githubusercontent.com/5788607/76036200-6f067b80-5f11-11ea-8936-0a2b4adcb7eb.png)

### Fiddle or Project
The stack overflow specifically comes on save from having added a node as a child to a text node. In this case and in the production code where I found this, the text node was returned by the previous sibling property where I was thinking that property would give me a usable HTML node. Note: the overflow only happens if the destination is otherwise saveable.
```csharp
string content = @"<html>

<head>
 <title>Test</title>
</head>
<body>
</body>
</html>

";

//load file
HtmlAgilityPack.HtmlDocument doc = new HtmlAgilityPack.HtmlDocument();
doc.LoadHtml(content);

doc.DocumentNode.SelectSingleNode("//head").PreviousSibling.AppendChild(HtmlNode.CreateNode(@"&nbsp;"));

doc.Save(@"C:\Users\jmessenger\Downloads\dummy2.htm");
```
### Further technical details
This happens on .Net 4.8 for sure and I believe 4.7 as well.

**Comment by JonathanMagnan** [comment]
Hello @jmessenger235 ,

My developer looked at this.

You currently try to append a node to a `text` node which is not allowed. If you were planning to take an element node, make sure to check that's not a text node if you want to `AppendChild`.

If you wish to add a text node, you might use the `InsertAfter` method or something like this:

```csharp
var previousSibling = doc.DocumentNode.SelectSingleNode("//head").PreviousSibling;
doc.DocumentNode.ChildNodes[0].InsertAfter(HtmlNode.CreateNode(@"&nbsp;"), previousSibling);
```

Or simply replace the current text.

Let me know if that helped.

Best Regards,

Jon

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Entity Framework Classic](http://entityframework-classic.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by jmessenger235** [comment]
Hi Jon,

Thank you for the code. As mentioned, I worked around it so I wasn't looking for a solution. The issue is that there's not an easy to understand error for this situation. A generic stack overflow on save doesn't seem related to adding a child to a text node. If it's not supported, it should throw an error on adding a child to a text node so that the developer can more quickly identify what is causing problems. In my example code, the issue is not very complicated to diagnose. In large HTML documents and code with complex DOM manipulations, identifying the problem took some time that could have been drastically reduced by a more relevant exception. AppendChild could simply add a check for type and throw an exception when the node having a child node added is a text or comment node. That way, any developer encountering a similar issue has an applicable exception at the point of the error instead of a generic stack overflow later.

Regards,
Joshua

**Comment by JonathanMagnan** [comment]
I understand,

If the library was new or barely used, we would definitely throw an error (we already did the code for this when testing).

But what we learned from `HAP` when doing modification on it is to always be backward compatible no matter what. There are so many people using the library that any changes will break some projects somewhere, so for this kind of case, we prefer doing nothing.
